### PR TITLE
Bg_Relay_Objects OK

### PR DESCRIPTION
**Pull request by Zelllll**
This is the doors used in the Dampe's grave minigame as well as the spinning gear in the Windmill Hut. All functions matching.

### CODE REVIEW

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 1]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+        DynaPolyInfo_Alloc(&D_060025FC, &sp24);
+        if (gSaveContext.eventChkInf[6] & 0x20) {
+            thisx->posRot.rot.y = 0x400;
+        } else {
+            thisx->posRot.rot.y = 0x80;
+        }
+        func_800F5718();
+        thisx->room = -1;
+        thisx->flags |= 0x20;
+        if ((D_808A9508 & 2) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:72]:
```suggestion
        if (D_808A9508 & 2) {
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 2]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+        if (this->unk_169 >= 6) {
+            if (D_808A9508 & 1) {
+                Actor_Kill(thisx);
+            } else {
+                D_808A9508 |= 1;
+                this->actionFunc = BgRelayObjects_Noop;
+            }
+        } else if (this->unk_169 != 5) {
+            Flags_UnsetSwitch(globalCtx, this->switchFlag);
+            if (((1 << this->unk_169) & D_808A9508)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:97]: I suggest writing like this if it works
```suggestion
            if (D_808A9508 & (1 << this->unk_169)) {
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 3]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+    if ((this->dyna.actor.params == WINDMILL_ROTATING_GEAR) && (gSaveContext.cutsceneIndex < 0xFFF0)) {
+        gSaveContext.eventChkInf[6] &= ~0x20;
+    }
+}
+
+void func_808A90F4(BgRelayObjects* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->switchFlag) != 0) {
+        if (this->timer != 0) {
+            Audio_PlayActorSound2(this, 0x2814);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:125]:
```suggestion
            Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_SLIDE_DOOR_OPEN);
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 4]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+    BgRelayObjects* this = THIS;
+
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+    if ((this->dyna.actor.params == WINDMILL_ROTATING_GEAR) && (gSaveContext.cutsceneIndex < 0xFFF0)) {
+        gSaveContext.eventChkInf[6] &= ~0x20;
+    }
+}
+
+void func_808A90F4(BgRelayObjects* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->switchFlag) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:123]:
```suggestion
    if (Flags_GetSwitch(globalCtx, this->switchFlag)) {
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 5]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+static InitChainEntry sInitChain[] = {
+    ICHAIN_F32(gravity, 5, ICHAIN_CONTINUE),
+    ICHAIN_VEC3F_DIV1000(scale, 100, ICHAIN_STOP),
+};
+
+void BgRelayObjects_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgRelayObjects* this = THIS;
+    s32 pad;
+    s32 sp24 = 0;
+    static u32 D_808A9508 = 0;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:56]:
```suggestion
    static u32 D_808A9508 = 0;
    BgRelayObjects* this = THIS;
    s32 pad;
    s32 sp24 = 0;
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 6]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+    if ((this->dyna.actor.params == WINDMILL_ROTATING_GEAR) && (gSaveContext.cutsceneIndex < 0xFFF0)) {
+        gSaveContext.eventChkInf[6] &= ~0x20;
+    }
+}
+
+void func_808A90F4(BgRelayObjects* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->switchFlag) != 0) {
+        if (this->timer != 0) {
+            Audio_PlayActorSound2(this, 0x2814);
+            if (gSaveContext.inventory.items[gItemSlots[10]] != 0xFF) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:126]:
```suggestion
            if (INV_CONTENT(ITEM_HOOKSHOT) != ITEM_NONE) {
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 7]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+void BgRelayObjects_Noop(BgRelayObjects* this, GlobalContext* globalCtx) {
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Relay_Objects/func_808A9234.s")
+void func_808A932C(BgRelayObjects* this, GlobalContext* globalCtx) {
+    if (this->timer != 0) {
+        this->timer--;
+    }
+    if (this->timer == 0) {
+        if (Player_InCsMode(globalCtx) == 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:179]:
```suggestion
        if (!Player_InCsMode(globalCtx)) {
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 8]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+            func_80078884(NA_SE_OC_ABYSS);
+            Gameplay_TriggerRespawn(globalCtx);
+            this->actionFunc = BgRelayObjects_Noop;
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Relay_Objects/func_808A9320.s")
+void func_808A939C(BgRelayObjects* this, GlobalContext* globalCtx) {
+    if (Flags_GetEnv(globalCtx, 5) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:188]:
```suggestion
    if (Flags_GetEnv(globalCtx, 5)) {
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 9]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+        gSaveContext.eventChkInf[6] = gSaveContext.eventChkInf[6] | 0x20;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:189]:
```suggestion
        gSaveContext.eventChkInf[6] |= 0x20;
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 10]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+    }
+    if ((gSaveContext.eventChkInf[6] & 0x20) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:191]:
```suggestion
    if (gSaveContext.eventChkInf[6] & 0x20) {
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 11]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+        Math_ApproxUpdateScaledS(&this->dyna.actor.posRot.rot.y, 0x400, 8);
+    } else {
+        Math_ApproxUpdateScaledS(&this->dyna.actor.posRot.rot.y, 0x80, 8);
+    }
+    this->dyna.actor.shape.rot.y = this->dyna.actor.shape.rot.y + this->dyna.actor.posRot.rot.y;
+    func_800F436C(&this->dyna.actor.projectedPos, 0x2076,
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:197]:
```suggestion
    func_800F436C(&this->dyna.actor.projectedPos, NA_SE_EV_WOOD_GEAR - SFX_FLAG,
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 12]
```diff
@@ -10,17 +10,28 @@
...
+void func_808A91AC(BgRelayObjects* this, GlobalContext* globalCtx);
+void func_808A9234(BgRelayObjects* this, GlobalContext* globalCtx);
+void BgRelayObjects_Noop(BgRelayObjects* this, GlobalContext* globalCtx);
+void func_808A932C(BgRelayObjects* this, GlobalContext* globalCtx);
+void func_808A939C(BgRelayObjects* this, GlobalContext* globalCtx);
+
 extern UNK_TYPE D_060001A0;
 extern UNK_TYPE D_060003C4;
 extern UNK_TYPE D_06001AB0;
 extern UNK_TYPE D_060025FC;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:33]: Based on their usage, the display lists can be typed correctly here:
```suggestion
extern Gfx D_060001A0[];
extern UNK_TYPE D_060003C4;
extern Gfx D_06001AB0[];
extern UNK_TYPE D_060025FC;
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 13]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+            if (((1 << this->unk_169) & D_808A9508)) {
+                Actor_Kill(thisx);
+            } else {
+                D_808A9508 |= (1 << this->unk_169);
+                this->actionFunc = func_808A90F4;
+            }
+        } else {
+            Flags_SetSwitch(globalCtx, this->switchFlag);
+            this->actionFunc = func_808A91AC;
+            thisx->posRot.pos.y = thisx->posRot.pos.y + 120.0f;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:106]:
```suggestion
            thisx->posRot.pos.y += 120.0f;
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 14]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+void func_808A939C(BgRelayObjects* this, GlobalContext* globalCtx) {
+    if (Flags_GetEnv(globalCtx, 5) != 0) {
+        gSaveContext.eventChkInf[6] = gSaveContext.eventChkInf[6] | 0x20;
+    }
+    if ((gSaveContext.eventChkInf[6] & 0x20) != 0) {
+        Math_ApproxUpdateScaledS(&this->dyna.actor.posRot.rot.y, 0x400, 8);
+    } else {
+        Math_ApproxUpdateScaledS(&this->dyna.actor.posRot.rot.y, 0x80, 8);
+    }
+    this->dyna.actor.shape.rot.y = this->dyna.actor.shape.rot.y + this->dyna.actor.posRot.rot.y;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:196]:
```suggestion
    this->dyna.actor.shape.rot.y += this->dyna.actor.posRot.rot.y;
```

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 15]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+        func_800AA000(this->dyna.actor.xyzDistFromLinkSq, 180, 20, 100);
+        Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_STONE_BOUND);
+        if (this->unk_169 != globalCtx->roomCtx.curRoom.num) {
+            func_800788CC(NA_SE_EN_PO_LAUGH);
+            this->timer = 5;
+            this->actionFunc = func_808A932C;
+            return;
+        }
+        Flags_UnsetSwitch(globalCtx, this->switchFlag);
+        this->dyna.actor.flags = this->dyna.actor.flags & ~0x10;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:163]:
```suggestion
        this->dyna.actor.flags &= ~0x10;
```

**Comment by Zelllll** [comment]
Thanks for all the suggestions, I'm surprised I missed that many! All are applied.

**Review by Roman971 — APPROVED** [review]

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 16]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Relay_Objects/BgRelayObjects_Update.s")
+void BgRelayObjects_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    BgRelayObjects* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Relay_Objects/BgRelayObjects_Draw.s")
+    if (this->dyna.actor.params == WINDMILL_ROTATING_GEAR) {
+        Gfx_DrawDListOpa(globalCtx, &D_06001AB0);
+    } else {
+        Gfx_DrawDListOpa(globalCtx, &D_060001A0);
```
Comment (fig02) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:213]: dont need & on these

File: src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c [hunk 17]
```diff
@@ -32,23 +43,173 @@ const ActorInit Bg_Relay_Objects_InitVars = {
...
+        if (globalCtx->roomCtx.curRoom.num == 4) {
+            gSaveContext.timer1State = 0xF;
+        }
+        this->actionFunc = BgRelayObjects_Noop;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Relay_Objects/func_808A91AC.s")
+void BgRelayObjects_Noop(BgRelayObjects* this, GlobalContext* globalCtx) {
+}
```
Comment (fig02) [inline src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:172]: I prefer DoNothing when describing an action, but we will discuss and standardize things like this at a later time. just mentioning it now

Comment (Zelllll) [reply src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:172]: yeah I'll change that, I thought Noop was kinda stupid tbh but I saw a lot of actors with it

Comment (fig02) [reply src/overlays/actors/ovl_Bg_Relay_Objects/z_bg_relay_objects.c:172]: its fine depending on context, in this case its an action where the actor is literally doing nothing

**Review by fig02 — APPROVED** [review]

**Pull request merged**